File: examples/python/agg-order/handlers/confirm_payment.py
"""Handler for ConfirmPayment command."""

from datetime import datetime, timezone

from google.protobuf.any_pb2 import Any
from google.protobuf.timestamp_pb2 import Timestamp

from angzarr import types_pb2 as types
from errors import CommandRejectedError
from proto import order_pb2 as order

from .state import OrderState


def handle_confirm_payment(
    command_book: types.CommandBook,
    command_any: Any,
    state: OrderState,
    seq: int,
) -> types.EventBook:
    """Handle ConfirmPayment command."""
    if not state.exists():
        raise CommandRejectedError("Order does not exist")
    if not state.is_payment_submitted():
        raise CommandRejectedError("Payment not submitted")

    cmd = order.ConfirmPayment()
    command_any.Unpack(cmd)

    if not cmd.payment_reference:
        raise CommandRejectedError("Payment reference is required")

    # 1 point per dollar
    loyalty_points_earned = state.total_after_discount() // 100

    event = order.OrderCompleted(
        final_total_cents=state.total_after_discount(),
        payment_method=state.payment_method,
        payment_reference=cmd.payment_reference,
        loyalty_points_earned=loyalty_points_earned,
        completed_at=Timestamp(seconds=int(datetime.now(timezone.utc).timestamp())),
        customer_root=state.customer_root,
        cart_root=state.cart_root,
    )
    event.items.extend(
        order.LineItem(
            product_id=i.product_id,
            name=i.name,
            quantity=i.quantity,
            unit_price_cents=i.unit_price_cents,
        )
        for i in state.items
    )

    event_any = Any()
    event_any.Pack(event, type_url_prefix="type.examples/")

    return types.EventBook(
        cover=command_book.cover,
        pages=[
            types.EventPage(
                num=seq,
                event=event_any,
                created_at=Timestamp(seconds=int(datetime.now(timezone.utc).timestamp())),
            )
        ],
    )

File: examples/rust/agg-order/src/handlers/confirm_payment.rs
//! Handler for ConfirmPayment command.

use angzarr::proto::{CommandBook, EventBook};
use common::proto::{ConfirmPayment, OrderCompleted, OrderState};
use common::{decode_command, now, require_exists, require_status_not, Result};

use crate::errmsg;
use crate::state::{calculate_total, state_builder};

/// Handle the ConfirmPayment command.
pub fn handle_confirm_payment(
    command_book: &CommandBook,
    command_data: &[u8],
    state: &OrderState,
    next_seq: u32,
) -> Result<EventBook> {
    require_exists(&state.customer_id, errmsg::ORDER_NOT_FOUND)?;
    require_status_not(&state.status, "pending", errmsg::PAYMENT_NOT_SUBMITTED)?;
    require_status_not(&state.status, "completed", errmsg::ORDER_COMPLETED)?;
    require_status_not(&state.status, "cancelled", errmsg::ORDER_CANCELLED)?;

    let cmd: ConfirmPayment = decode_command(command_data)?;

    let final_total = calculate_total(state);
    // 1 loyalty point per $1 (100 cents)
    let loyalty_points_earned = final_total / 100;

    let event = OrderCompleted {
        final_total_cents: final_total,
        payment_method: state.payment_method.clone(),
        payment_reference: cmd.payment_reference.clone(),
        loyalty_points_earned,
        completed_at: Some(now()),
        customer_root: state.customer_root.clone(),
        cart_root: state.cart_root.clone(),
        items: state.items.clone(),
        // Default to approved when no fraud service configured (router path)
        fraud_check_result: "approved".to_string(),
    };

    Ok(state_builder().build_response(
        state,
        command_book.cover.clone(),
        next_seq,
        "type.examples/examples.OrderCompleted",
        event,
    ))
}
